File: app/src/main/java/com/kickstarter/services/apiresponses/MessageThreadEnvelope.java
package com.kickstarter.services.apiresponses;

import android.os.Parcelable;

import com.kickstarter.libs.qualifiers.AutoGson;
import com.kickstarter.models.Message;
import com.kickstarter.models.MessageThread;
import com.kickstarter.models.User;

import java.util.List;

import androidx.annotation.Nullable;
import auto.parcel.AutoParcel;

@AutoGson
@AutoParcel
public abstract class MessageThreadEnvelope implements Parcelable {
  public abstract @Nullable List<Message> messages();
  public abstract @Nullable MessageThread messageThread();
  public abstract List<User> participants();

  @AutoParcel.Builder
  public abstract static class Builder {
    public abstract Builder messages(List<Message> __);
    public abstract Builder messageThread(MessageThread __);
    public abstract Builder participants(List<User> __);
    public abstract MessageThreadEnvelope build();
  }

  public static Builder builder() {
    return new AutoParcel_MessageThreadEnvelope.Builder();
  }

  public abstract Builder toBuilder();
}

File: app/src/main/java/com/kickstarter/services/apiresponses/MessageThreadEnvelope.kt
package com.kickstarter.services.apiresponses

import android.os.Parcelable
import com.kickstarter.models.Message
import com.kickstarter.models.MessageThread
import com.kickstarter.models.User
import kotlinx.parcelize.Parcelize

@Parcelize
class MessageThreadEnvelope private constructor(
    private val messages: List<Message> = emptyList(),
    private val messageThread: MessageThread?,
    private val participants: List<User> = emptyList(),
) : Parcelable {
    fun messages() = this.messages
    fun messageThread() = this.messageThread
    fun participants() = this.participants

    @Parcelize
    data class Builder(
        private var messages: List<Message> = emptyList(),
        private var messageThread: MessageThread? = null,
        private var participants: List<User> = emptyList()
    ) : Parcelable {
        fun messages(messages: List<Message>?) = apply { this.messages = messages ?: emptyList() }
        fun messageThread(messageThread: MessageThread?) = apply { this.messageThread = messageThread }
        fun participants(participants: List<User>) = apply { this.participants = participants }
        fun build() = MessageThreadEnvelope(
            messages = messages,
            messageThread = messageThread,
            participants = participants
        )
    }

    override fun equals(obj: Any?): Boolean {
        var equals = super.equals(obj)
        if (obj is MessageThreadEnvelope) {
            equals = messages() == obj.messages() &&
                messageThread() == obj.messageThread() &&
                participants() == obj.participants()
        }
        return equals
    }

    fun toBuilder() = Builder(
        messages = messages,
        messageThread = messageThread,
        participants = participants
    )

    companion object {
        @JvmStatic
        fun builder() = Builder()
    }
}
